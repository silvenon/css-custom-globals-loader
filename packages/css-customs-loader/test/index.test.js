const requireFromString = require('require-from-string')
const path = require('path')
const compile = require('./compile')

const cssCustomsLoader = require.resolve('../')
const getCompiledExportedValue = ({ stats, entry }) => {
  const { source } = stats
    .toJson()
    .modules.find(
      ({ name, identifier }) =>
        name.includes(entry) && identifier.includes(cssCustomsLoader)
    )
  return requireFromString(
    source,
    `${__dirname}/fixtures/${path.basename(entry)}`
  )
}

describe(`emits an error`, () => {
  test(`when css-customs-loader is placed after css-loader`, async () => {
    await expect(
      compile({
        entry: './fixtures/basic.css',
        rules: [
          {
            test: /\.css$/,
            use: [
              {
                loader: 'css-loader',
                options: {
                  importLoaders: 2,
                },
              },
              cssCustomsLoader,
              'postcss-loader',
            ],
          },
        ],
      })
    ).rejects.toMatchSnapshot()
  })

  test(`when postcss-preset-env is missing`, async () => {
    await expect(
      compile({
        entry: './fixtures/basic.css',
        rules: [
          {
            test: /\.css$/,
            use: [
              cssCustomsLoader,
              {
                loader: 'css-loader',
                options: {
                  importLoaders: 1,
                },
              },
              {
                loader: 'postcss-loader',
                options: {
                  plugins: () => [require('postcss-brand-colors')],
                },
              },
            ],
          },
        ],
      })
    ).rejects.toMatchSnapshot()
  })

  test(`when postcss-loader is missing`, async () => {
    await expect(
      compile({
        entry: './fixtures/basic.css',
        rules: [
          {
            test: /\.css$/,
            use: [cssCustomsLoader, 'css-loader'],
          },
        ],
      })
    ).rejects.toMatchSnapshot()
  })
})

it(`exposes CSS customs in the default export object`, async () => {
  const { stats, entry } = await compile({
    entry: './fixtures/basic.css',
    rules: [
      {
        test: /\.css$/,
        use: [
          cssCustomsLoader,
          {
            loader: 'css-loader',
            options: {
              importLoaders: 1,
            },
          },
          'postcss-loader',
        ],
      },
    ],
  })
  const { locals } = getCompiledExportedValue({ stats, entry })
  expect(locals).toMatchInlineSnapshot(`
    Object {
      "customMedia": Object {
        "--narrow-window": "(max-width: 30em)",
      },
      "customProperties": Object {
        "--primary-color": "lightblue",
      },
      "customSelectors": Object {
        ":--title": "h1",
      },
    }
  `)
})

it(`exposes CSS Modules in the same object as customs`, async () => {
  const { stats, entry } = await compile({
    entry: './fixtures/modules.css',
    rules: [
      {
        test: /\.css$/,
        use: [
          cssCustomsLoader,
          {
            loader: 'css-loader',
            options: {
              importLoaders: 1,
              modules: true,
            },
          },
          'postcss-loader',
        ],
      },
    ],
  })
  const { locals } = getCompiledExportedValue({ stats, entry })
  expect(locals).toMatchInlineSnapshot(`
    Object {
      "customMedia": Object {
        "--narrow-window": "(max-width: 30em)",
      },
      "customProperties": Object {
        "--primary-color": "lightblue",
      },
      "customSelectors": Object {
        ":--title": "h1",
      },
      "text": "_3KySKA6rf07sl0ilq7bBBm",
    }
  `)
})

it(`can export only locals`, async () => {
  const { stats, entry } = await compile({
    entry: './fixtures/modules.css',
    rules: [
      {
        test: /\.css$/,
        use: [
          `${cssCustomsLoader}?onlyLocals`,
          {
            loader: 'css-loader',
            options: {
              importLoaders: 1,
              modules: true,
              onlyLocals: true,
            },
          },
          'postcss-loader',
        ],
      },
    ],
  })
  const result = getCompiledExportedValue({ stats, entry })
  expect(result).toMatchInlineSnapshot(`
    Object {
      "customMedia": Object {
        "--narrow-window": "(max-width: 30em)",
      },
      "customProperties": Object {
        "--primary-color": "lightblue",
      },
      "customSelectors": Object {
        ":--title": "h1",
      },
      "text": "_3KySKA6rf07sl0ilq7bBBm",
    }
  `)
})

it(`supports files with external @imports`, async () => {
  await expect(
    compile({
      entry: './fixtures/external.css',
      rules: [
        {
          test: /\.css$/,
          include: `${__dirname}/fixtures`,
          use: [
            cssCustomsLoader,
            {
              loader: 'css-loader',
              options: {
                importLoaders: 1,
              },
            },
            {
              loader: 'postcss-loader',
              options: {
                config: { path: __dirname },
              },
            },
          ],
        },
      ],
    })
  ).resolves.toBeDefined()
})

it(`uses PostCSS plugins before postcss-preset-env`, async () => {
  const { stats, entry } = await compile({
    entry: './fixtures/postcss-plugins.css',
    rules: [
      {
        test: /\.css$/,
        use: [
          cssCustomsLoader,
          {
            loader: 'css-loader',
            options: {
              importLoaders: 1,
            },
          },
          {
            loader: 'postcss-loader',
            options: {
              plugins: () => [
                require('postcss-brand-colors'),
                require('postcss-preset-env'),
              ],
            },
          },
        ],
      },
    ],
  })
  const { locals } = getCompiledExportedValue({ stats, entry })
  expect(locals).toMatchInlineSnapshot(`
    Object {
      "customMedia": Object {},
      "customProperties": Object {
        "--primary-color": "#1da1f2",
      },
      "customSelectors": Object {},
    }
  `)
})

it('uses webpack loaders after postcss-loader', async () => {
  const { stats, entry } = await compile({
    entry: './fixtures/loaders.less',
    rules: [
      {
        test: /\.less$/,
        use: [
          cssCustomsLoader,
          {
            loader: 'css-loader',
            options: {
              importLoaders: 2,
            },
          },
          'postcss-loader',
          'less-loader',
        ],
      },
    ],
  })
  const { locals } = getCompiledExportedValue({ stats, entry })
  expect(locals).toMatchInlineSnapshot(`
    Object {
      "customMedia": Object {},
      "customProperties": Object {
        "--primary-color": "lightblue",
      },
      "customSelectors": Object {},
    }
  `)
})
