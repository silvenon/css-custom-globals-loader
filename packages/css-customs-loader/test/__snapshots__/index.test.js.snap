// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emits an error when css-customs-loader is placed after css-loader 1`] = `
"./fixtures/basic.css
Module build failed (from ../index.js):
Error: css-customs-loader should be added BEFORE css-loader. <CWD>/node_modules/css-loader/dist/cjs.js??ref--4-0!<CWD>/packages/css-customs-loader/index.js!<CWD>/node_modules/postcss-loader/src/index.js!<CWD>/packages/css-customs-loader/test/fixtures/basic.css
    at Object.exports.addBeforeCssLoader (<CWD>/packages/css-customs-loader/lib/errors.js:2:3)
    at Object.addBeforeCssLoader (<CWD>/packages/css-customs-loader/lib/index.js:21:27)"
`;

exports[`emits an error when postcss-loader is missing 1`] = `
"./fixtures/basic.css
Module build failed (from ../index.js):
Error: 
postcss-loader is missing, you need to use it in order for css-customs-loader to work properly.
  
    at Object.<anonymous> (<CWD>/packages/css-customs-loader/lib/errors.js:8:32)
    at Runtime._execModule (<CWD>/node_modules/jest-runtime/build/index.js:1034:22)
    at Runtime._loadModule (<CWD>/node_modules/jest-runtime/build/index.js:678:12)
    at Runtime.requireModule (<CWD>/node_modules/jest-runtime/build/index.js:536:10)
    at Runtime.requireModuleOrMock (<CWD>/node_modules/jest-runtime/build/index.js:699:21)
    at Object.require (<CWD>/packages/css-customs-loader/lib/index.js:9:15)
    at Runtime._execModule (<CWD>/node_modules/jest-runtime/build/index.js:1034:22)
    at Runtime._loadModule (<CWD>/node_modules/jest-runtime/build/index.js:678:12)
    at Runtime.requireModule (<CWD>/node_modules/jest-runtime/build/index.js:536:10)
    at Runtime.requireModuleOrMock (<CWD>/node_modules/jest-runtime/build/index.js:699:21)"
`;

exports[`emits an error when postcss-preset-env is missing 1`] = `
"./fixtures/basic.css
Module build failed (from ../index.js):
Error: 
postcss-preset-env is missing from your plugins, you need to use it in order for css-customs-loader to work properly.
  
    at Object.<anonymous> (<CWD>/packages/css-customs-loader/lib/errors.js:14:35)
    at Runtime._execModule (<CWD>/node_modules/jest-runtime/build/index.js:1034:22)
    at Runtime._loadModule (<CWD>/node_modules/jest-runtime/build/index.js:678:12)
    at Runtime.requireModule (<CWD>/node_modules/jest-runtime/build/index.js:536:10)
    at Runtime.requireModuleOrMock (<CWD>/node_modules/jest-runtime/build/index.js:699:21)
    at Object.require (<CWD>/packages/css-customs-loader/lib/index.js:9:15)
    at Runtime._execModule (<CWD>/node_modules/jest-runtime/build/index.js:1034:22)
    at Runtime._loadModule (<CWD>/node_modules/jest-runtime/build/index.js:678:12)
    at Runtime.requireModule (<CWD>/node_modules/jest-runtime/build/index.js:536:10)
    at Runtime.requireModuleOrMock (<CWD>/node_modules/jest-runtime/build/index.js:699:21)"
`;
